android/app/src/main/java/com/microsoft/codepush/common/core/CodePushBaseCore.java

```diff
@@ -0,0 +1,1043 @@
+                } else {
+                    AppCenterLog.info(CodePush.LOG_TAG, "Reporting CodePush update rollback (" + statusReport.getLabel() + ")");
+                }
+                configuration.setDeploymentKey(statusReport.getPackage().getDeploymentKey());
+                mManagers.mAcquisitionManager.reportStatusDeploy(configuration, statusReport);
+                saveReportedStatus(statusReport);
+            }
+        } catch (CodePushReportStatusException e) {
+
+            /* In order to do not lose original exception if another one will be thrown during the retry
+             * we need to wrap it */
+            CodePushNativeApiCallException exceptionToThrow = new CodePushNativeApiCallException(e);
+            try {
+                retrySendStatusReport(statusReport);
+            } catch (CodePushNativeApiCallException retryException) {
+                exceptionToThrow = new CodePushNativeApiCallException(exceptionToThrow);
+            }
+            throw exceptionToThrow;
+        }
+
+        /* If there was several attempts to send error reports */
+        clearScheduledAttemptsToRetrySendStatusReport();
+    }
+
+    /**
+     * Attempts to retry sending status report if there was sending error before.
+     *
+     * @param statusReport status report.
+     * @throws CodePushNativeApiCallException if error occurred during the execution of operation.
+     */
+    private void retrySendStatusReport(CodePushDeploymentStatusReport statusReport) throws CodePushNativeApiCallException {
+
+        /* Try again when the app resumes */
+        /* TODO check that statusReport.toString() will be serialized into JSON string! */
+        AppCenterLog.info(CodePush.LOG_TAG, "Report status failed: " + statusReport.toString());
+        saveStatusReportForRetry(statusReport);
+        Callable<Void> sender = new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+                final CodePushDeploymentStatusReport statusReport = getNewStatusReport();
+                if (statusReport != null) {
+                    tryReportStatus(statusReport);
+                }
+                return null;
+            }
+        };
+        try {
+            retrySendStatusReportOnAppResume(sender);
+        } catch (Exception e) {
+            throw new CodePushNativeApiCallException("Error retry sending status report. ", e);
+        }
+    }
+
+    /**
+     * Retries to send status report on app resume using platform-specific way for it.
+     * Use <code>sender.call()</code> to invoke sending of report.
+     *
+     * @param sender task that sends status report.
+     * @throws Exception if error occurred during the process.
+     */
+    @SuppressWarnings("WeakerAccess")
+    protected abstract void retrySendStatusReportOnAppResume(Callable<Void> sender) throws Exception;
```

Comment:
Put abstract methods either on top or at the bottom of the class. I think strict order better be used.

Reply:
@annakocheshkova agree.